pkg/applicationserver/applicationserver.go

```diff
@@ -253,47 +238,69 @@ func (as *ApplicationServer) Roles() []ttnpb.ClusterRole {
+// traffic of the application. Otherwise, only traffic that was processed locally is sent.
+func (as *ApplicationServer) Subscribe(ctx context.Context, protocol string, ids *ttnpb.ApplicationIdentifiers, cluster bool) (*io.Subscription, error) {
+	ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("as:conn:%s", events.NewCorrelationID()))
+	if ids != nil {
+		uid := unique.ID(ctx, ids)
+		ctx = log.NewContextWithField(ctx, "application_uid", uid)
 	}
+	if cluster {
+		return as.clusterDistributor.Subscribe(ctx, protocol, ids)
+	} else {
+		return as.localDistributor.Subscribe(ctx, protocol, ids)
+	}
+}
 
-	uid := unique.ID(ctx, ids)
-	ctx = log.NewContextWithField(
-		events.ContextWithCorrelationID(ctx, fmt.Sprintf("as:conn:%s", events.NewCorrelationID())),
-		"application_uid", uid,
-	)
+// SendUp processes the given upstream message and then broadcasts it to the application frontends.
+func (as *ApplicationServer) SendUp(ctx context.Context, up *ttnpb.ApplicationUp) error {
+	link, err := as.linkRegistry.Get(ctx, up.ApplicationIdentifiers, []string{
+		"default_formatters",
+		"skip_payload_crypto",
+	})
+	if err != nil {
+		return err
+	}
+	return as.processUp(ctx, up, link)
+}
+
+func (as *ApplicationServer) processUp(ctx context.Context, up *ttnpb.ApplicationUp, link *ttnpb.ApplicationLink) error {
+	ctx = events.ContextWithCorrelationID(ctx, append(up.CorrelationIDs, fmt.Sprintf("as:up:%s", events.NewCorrelationID()))...)
+	up.CorrelationIDs = events.CorrelationIDsFromContext(ctx)
+	registerReceiveUp(ctx, up)
+
+	now := time.Now().UTC()
+	up.ReceivedAt = &now
 
-	link, err := as.getLink(ctx, ids)
+	pass, err := as.handleUp(ctx, up, link)
 	if err != nil {
-		return nil, err
+		log.FromContext(ctx).WithError(err).Warn("Failed to process upstream message")
+		registerDropUp(ctx, up, err)
+		return nil
 	}
-	sub := io.NewSubscription(ctx, protocol, &ids)
-	select {
-	case <-link.ctx.Done():
-		return nil, link.ctx.Err()
-	case link.subscribeCh <- sub:
-	}
-	go func() {
-		select {
-		case <-link.ctx.Done():
-			// Disconnect the subscription in order to avoid leaking it,
-			// and skip the unsubscribe channel since it will get closed.
-			sub.Disconnect(link.ctx.Err())
-			return
-		case <-sub.Context().Done():
-		}
-		select {
-		case <-link.ctx.Done():
-			return
-		case link.unsubscribeCh <- sub:
-		}
-	}()
-	return sub, nil
+	if !pass {
+		return nil
+	}
+
+	if err := as.broadcastUp(ctx, up); err != nil {
+		log.FromContext(ctx).WithError(err).Warn("Failed to broadcast upstream message")
+		registerDropUp(ctx, up, err)
+		return nil
+	}
+	registerForwardUp(ctx, up)
+
+	return nil
 }
 
-func skipPayloadCrypto(link *link, dev *ttnpb.EndDevice) bool {
+func (as *ApplicationServer) broadcastUp(ctx context.Context, up *ttnpb.ApplicationUp) error {
+	if err := as.localDistributor.SendUp(ctx, up); err != nil {
+		return err
+	}
+	return as.clusterDistributor.SendUp(ctx, up)
```

Comment:
Are they sequential for a reason? Are these calls blocking? If so, consider goroutines and a wait group

Reply:
Error handling is better if they are sequential, and both just write to a buffered channel, so I'm more inclined to leave it as is.

Reply:
I'm okay with optimizing this later.